idx-kotlin/src/main/java/com/okta/idx/kotlin/client/IdxClient.kt

```diff
@@ -211,33 +185,74 @@ class IdxClient internal constructor(
                 .build()
         }
 
-        return withContext(configuration.ioDispatcher) {
-            try {
-                val response = configuration.performRequest(request)
-                val tokenResponse =
-                    configuration.json.decodeFromString<Token>(response.body!!.string())
-
-                IdxClientResult.Response(
-                    TokenResponse(
-                        accessToken = tokenResponse.accessToken,
-                        expiresIn = tokenResponse.expiresIn,
-                        refreshToken = tokenResponse.refreshToken,
-                        idToken = tokenResponse.idToken,
-                        scope = tokenResponse.scope,
-                        tokenType = tokenResponse.tokenType,
-                    )
-                )
-            } catch (e: Exception) {
-                IdxClientResult.Error(e)
-            }
-        }
+        return configuration.performRequest(request, Token::toIdxResponse)
     }
 
     /**
      * Evaluates the given redirect url to determine what next steps can be performed. This is usually used when receiving a redirection from an IDP authentication flow.
      */
-    suspend fun redirectResult(url: HttpUrl) {
-        TODO()
+    suspend fun redirectResult(uri: Uri): IdxRedirectResult {
+        val errorQueryParameter = uri.getQueryParameter("error")
+        val stateQueryParameter = uri.getQueryParameter("state")
+        if (errorQueryParameter == "interaction_required") {
+            // Validate the state matches. This is a security assurance.
+            if (clientContext.state != stateQueryParameter) {
+                val error = "IDP redirect failed due to state mismatch."
+                return IdxRedirectResult.Error(error)
+            }
+            return when (val resumeResult = resume()) {
+                is IdxClientResult.Error -> {
+                    IdxRedirectResult.Error("Failed to resume.", resumeResult.exception)
+                }
+                is IdxClientResult.Response -> {
+                    IdxRedirectResult.InteractionRequired(resumeResult.response)
+                }
+            }
+        }
+        if (errorQueryParameter != null) {
+            val errorDescription =
+                uri.getQueryParameter("error_description") ?: "An error occurred."
+            return IdxRedirectResult.Error(errorDescription)
+        }
+        val interactionCodeQueryParameter = uri.getQueryParameter("interaction_code")
+        if (interactionCodeQueryParameter != null) {
+            // Validate the state matches. This is a security assurance.
+            if (clientContext.state != stateQueryParameter) {
+                val error = "IDP redirect failed due to state mismatch."
+                return IdxRedirectResult.Error(error)
+            }
+            return when (val result = exchangeCodes(interactionCodeQueryParameter)) {
```

Comment:
From reading this, it looks like the redirectResult will automatically attempt to exchange the code, am I right?

Reply:
Yes

Reply:
I was thinking this function wouldn't be async, and that it could be a way for a developer to see what they should do with a URL.  Think of it this way, an app may be launched / interacted with using a variety of URLs.  So my thought was that a developer could call this to ask "Is this relevant to IDX?  No?  Okay, I'll try something else".

That's why I separated out finding out what the result of a redirect is, and performing the next step / exchanging codes.

Reply:
I see, I think we could add a little preamble to this that eagerly rejects urls which don't match the `redirectUrl`.

I'll add this to my list of offline discussions/direction for future additions/refactors.